Add map block

andysellick opened the pull request:
, [Jira issue PNP-6356](https://gov-uk.atlassian.net/browse/PNP-6356)⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What
Adds a custom map block to the `landing_page` structure.

Partly based on initial work in https://github.com/alphagov/frontend/pull/4748

Tested in latest Chrome, Edge, Firefox, Opera, plus iPhone 14, Galaxy S24, Google Pixel 9, iPhone 12.

### Accessibility
I've added some additional commits to the end of this PR to address some accessibility issues that were found during a recent test. Making a note of some unsolved ones here and the reason.

- `Map markers missing discernable text` - when highlighted with a screen reader, there's no way of distinguishing one map marker from another. Ideally this would be possible by adding some kind of label but the JS that controls this is handled by Leaflet, and there appears to be no option in the Leaflet API to modify this.
- `Information in map isn't provided elsewhere` - this is a known issue relating to content, which we have tried to resolve.

## Why

## Visual changes

Trello card: https://trello.com/c/zVqSZM6D

andysellick committed:
Add initial placeholder block

andysellick committed:
Add initial block stylesheet

andysellick committed:
Add leaflet dependency

andysellick committed:
Add required stylesheets

andysellick committed:
Add required JavaScript

andysellick committed:
Use checkbox component for filter

andysellick committed:
Add initial analytics

- tracks filter changes and popup opening
- not fully correct yet, text attribute for popups isn't coming through properly, but may be related to wider problems with the map code that should be fixed in the next iteration

andysellick committed:
Improve FE and update to latest code

- updates the JS for the map to the latest code, including a few adjustments
- corresponds to v5 of the map
- removes cluster feature plugin in favour of a manual solution
- update bindpopup function significantly
- remove os-api-branding as no longer needed
- analytics for text attribute still not working correctly
- move JS rendering of copyright information into the template
- add styles to improve layout and positioning of map
- extend and improve key, including new icon for additional key item and change of content for key descriptions
- prevent map from appearing using CSS if JS is disabled

andysellick committed:
Add data

andysellick committed:
Fix uncoloured region on map

- the map uses a mask around the UK to hide other countries, but the image for this mask won't load, producing a pale red box around the edge of the map
- setting the background colour of the map element to the same colour as the ocean hides this problem

andysellick committed:
Clean up stylesheet indentation

andysellick committed:
Add map block documentation

andysellick committed:
Increase size of map popup close icon

andysellick committed:
Fix final bit of tracking

- had to get the title of the popup on popup open
- can't store a value in the ga4 object, so having to read the popup contents and look for the title using an attribute

andysellick committed:
Convert map code into a GOV.UK JS module

- remove old map code and add new module and test file
- add delayed check for cookie consent with analytics
- add styling to make map error state stand out a little more

KludgeKML commented on app/assets/stylesheets/views/_landing_page/map/leaflet.scss:
```diff
@@ -0,0 +1,752 @@
+/* stylelint-disable */
```
Can we not get leaflet's css from the included node module, or is this custom stuff?

andysellick replied:
Yes, you're right we can. Have updated (there are a few overrides that are needed, I've moved them across to the main map sass file).

KludgeKML reviewed (commented):
Looks okay to me, although ultimately I think you'd want to tidy up the commits to show a bit less of the process :) Can't comment on the ultimate conversion into a govuk JS module - I'll leave that to @AshGDS but it looks solid enough?

AshGDS commented on app/assets/images/missions/layers-2x.png:
I'm not sure you need all these images anymore - they were needed for v1 where there was a filter toggle on the map itself.

andysellick replied:
Thanks, have removed 👍

AshGDS reviewed (commented):
Looks good overall 👍 Just a bit concerned about the API key implementation though. 

Also probably a content (`.yml`) thing, but the page title was `GOV.UK` for me, and WAVE complains about no page language being specified.

andysellick committed:
Improve maps on mobile

- reduce spacing on mobile for popups, prevent collision with headings and close icons
- set maximum width of popups to prevent overlap on mobile
- set maximum height of map on mobile to not be more than screen height, to help with scrolling
- remove in-house leaflet CSS and use one from package instead, move overrides into main stylesheet

andysellick commented on app/views/landing_page/blocks/_map.html.erb:
```diff
@@ -1,5 +1,13 @@
-map
+<div id="api-warning"></div>
+
+<div class="map" id="map" data-api-key="<%= ENV["OS_MAPS_API_KEY"] %>"></div>
```
There is an option to build a proxy service to mask this key but we decided the effort involved was too great. We have the option instead to lock the key to only specific URLs, which might cause problems for development (I haven't looked into the specifics yet) but should certainly done after this is live.

andysellick commented:
@KludgeKML have had a go at squashing some of the commits, not got far but hopefully it's an improvement.

andysellick committed:
Remove popup icon active state

- this bit of the code changed the icons for popups to a slightly different appearance when the popup is opened
- however the lines of code that control this behaviour are quite buried in the code and extremely hard to provide test coverage for, which caused the test coverage to be below the required level
- additionally the visual change with these icons is very subtle and provides little benefit, particularly since the popup appearing is a highly visual indication that something has changed

andysellick committed:
Increase test coverage

- employ 'istanbul ignore next' to improve test coverage for the last few unreachable statements (including one that's definitely being tested, which I find confusing)
- rename some variables to be clearer
- change checkbox name attribute to be better (required, but not in use)

AshGDS reviewed (commented):
Looks good to me, can probably be approved purely looking at the code but I haven't looked at it locally yet.

Had some non-blocking comments - there's a chance you've already addressed them as I went commit by commit when reviewing so something I commented on might have changed in a future commit.

andysellick commented on app/assets/javascripts/modules/landing-page-map.js:
```diff
@@ -15,12 +15,10 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
         },
         hub: {
-          default: L.icon({ iconUrl: this.$module.getAttribute('data-icon-sh-default'), iconSize: iconSize, iconAnchor: [0, 12], popupAnchor: [12, 0] }),
-          active: L.icon({ iconUrl: this.$module.getAttribute('data-icon-sh-default'), iconSize: iconSize, iconAnchor: [0, 12], popupAnchor: [12, 0] })
```
Potentially, but I've since removed the selected state entirely as it was a blocker to increasing test coverage and I don't think actually added any benefit - the visual change of a popup appearing is plenty.

andysellick commented on app/assets/javascripts/views/landing_page/map/main.js:
```diff
@@ -164,18 +171,31 @@ window.addEventListener('DOMContentLoaded', function () {
       }
+
+      // track the filter click
+      mapElement.setAttribute('data-ga4-auto', JSON.stringify({
```
Good question. I think it's because we need `index_link` to be different for each checkbox, and I couldn't see a way of doing that without further code - unless there is a way to do that?

andysellick commented on app/assets/javascripts/views/landing_page/map/main.js:
```diff
@@ -190,121 +159,87 @@ window.addEventListener('DOMContentLoaded', function () {
-    properties = (({ ICB23CD, NHSER22CD, ...o }) => o)(properties)
+    const ftProps = layer.feature.properties
+    const ftGeomType = layer.feature.geometry.type
+    const lyrPane = layer.options.pane
```
No, good point. Have written out to be clearer. This was part of the code we received, there's probably a few things like this that I might have overlooked.

andysellick commented on app/assets/javascripts/views/landing_page/map/main.js:
```diff
@@ -178,6 +178,7 @@ window.addEventListener('DOMContentLoaded', function () {
 
     const popupHeading = document.createElement('h3')
     popupHeading.classList.add('govuk-heading-m')
+    popupHeading.setAttribute('data-ga4-text', '')
```
Good shout. This was unnecessarily overcomplicated anyway - I just needed a hook to get the element's text, and the class suits perfectly well for that. Updated.

AshGDS replied:
Nice, thanks!

andysellick commented on spec/javascripts/unit/modules/landing-page-map.spec.js:
```diff
@@ -0,0 +1,289 @@
+      expect(attributes).toEqual(JSON.stringify(expected))
+    })
+
+    // this test is a bit brittle as it depends on specific data, but at least it's tested
```
I initially thought I could do this by getting the thing I was about to click on and modifying the markup within it, but unfortunately it's a bit more complicated. The popups aren't connected to the markers in the DOM, they're a single thing in a separate layer. I could manually update that layer before clicking on the marker, but that change would immediately be overwritten by the map JS, and if I tried to update it after the click the event would already have happened.

Happy to hear other ideas, but given the nature of this thing I think I'd like to get it over the line and address less than good things like this later if it becomes a problem.

andysellick committed:
Fix error in the data

- two regions were incorrectly the wrong way round

andysellick committed:
Accessibility fix: skip map

- adds a skip link so keyboard and screen reader users can bypass the map quickly
- normally would aim for the next piece of content but because this is a block in a page of potentially unknown other blocks we can't specifically target what comes next, so have had to add a blank div at the end of the map block as the target
- this is a bit confusing for screen reader users as the target is then announced as an empty group, but probably better this than no skip functionality

andysellick committed:
Accessibility fix - map marker focus

- improve the visual appearance of map markers on focus
- add code so that partially obscured markers are brought to the front when mouse hovered
- sadly this can't be 100% effective as the markers are two different layers with specific z-indexes (and draw orders) so pink icons will always appear overlaid on blue ones, but can't fix that without merging those two layers then completely rewriting the filter checkbox code

andysellick committed:
Accessibility fix - change key title

- changes title of map key from 'Key' to 'Key for map' to improve clarity when read out of context

andysellick committed:
Accessibility fix - esc key closes popups

- bind the escape key so that when pressed map popups are closed

KludgeKML reviewed (approved):
LGTM

AshGDS commented on package.json:
```diff
@@ -51,5 +51,8 @@
-  "packageManager": "yarn@3.5.0"
+  "packageManager": "yarn@3.5.0",
+  "dependencies": {
+    "leaflet": "^1.9.4"
```
Just a thought, would dependabot pick up any leaflet upgrades and auto merge them? Do we need to keep an eye on that?

AshGDS commented on spec/javascripts/unit/modules/landing-page-map.spec.js:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    it('does nothing if there is no API key', function () {
+      setupMap('') // need to pass an empty string, otherwise becomes the string 'undefined'
```
You could have a default value in the function definition instead of passing an empty string. E.g. `function setupMap (apiKey = '') {`

andysellick replied:
Good shout, fixed. Also realised for the other version of this function I don't need to keep passing a fake apiKey as it never changes in that test.

AshGDS reviewed (approved):
LGTM 👍
